src/capi_flash_semptian_user.c

```diff
@@ -0,0 +1,445 @@
+
+#define CHECK(X) (assert((X) >= 0))
+
+int main (int argc, char *argv[])
+{
+  // int priv1,priv2;
+  int dat, dif, edat;
+  int CFG;
+  int RBF;
+  time_t ct, lt, st, et, eet, set, ept, spt, svt, evt;
+  int address, raddress;
+
+  char rbf_file[1024];
+  char cfg_file[1024];
+
+  int  print_cnt = 0;
+
+  if (argc < 3) {
+    printf("Usage: capi_flash <rbf_file> <card#>\n\n");
+    exit(-1);
+  }
+  strcpy (rbf_file, argv[1]);
+
+  if ((RBF = open(rbf_file, O_RDONLY)) < 0) {
+    printf("Can not open %s\n",rbf_file);
+    exit(-1);
+  }
+
+  strcpy(cfg_file, "/sys/class/cxl/card");
+  strcat(cfg_file, argv[2]);
+  strcat(cfg_file, "/device/config");
+
+  if ((CFG = open(cfg_file, O_RDWR)) < 0) {
+    printf("Can not open %s\n",cfg_file);
+    exit(-1);
+  }
+
+  int temp,vendor,device, subsys;
+  lseek(CFG, 0, SEEK_SET);
+  CHECK( read(CFG, &temp, 4) );
+  vendor = temp & 0xFFFF;
+  device = (temp >> 16) & 0xFFFF;  
+  printf("Device ID: %04X\n", device);
+  printf("Vendor ID: %04X\n", vendor);
+
+  lseek(CFG, 44, SEEK_SET);
+  CHECK( read(CFG, &temp, 4) );
+  subsys = (temp >> 16) & 0xFFFF;
+
+  if ( (vendor != 0x1014) || ( device != 0x0477) || ((subsys != 0x0607) && (subsys != 0x0606))) {
+	  printf("Not the correct flashing binary for card\n");
+	  exit(-1);
+  }
+  
+  int addr_reg, size_reg, cntl_reg, data_reg;
+  lseek(CFG, 0x404, SEEK_SET);
+  CHECK( read(CFG, &temp,4) );
+  printf("  VSEC Length/VSEC Rev/VSEC ID: 0x%08X\n", temp);
+  if ( (temp & 0x08000000) == 0x08000000 ) {
+    printf("    Version 0.12\n\n");
+    addr_reg = 0x450;
+    size_reg = 0x454;
+    cntl_reg = 0x458;
+    data_reg = 0x45c;
+  } else {
+    printf("    Version 0.10\n\n");
+    addr_reg = 0x920;
+    size_reg = 0x924;
+    cntl_reg = 0x928;
+    data_reg = 0x92c;
+  }
+ 
+  // Set stdout to autoflush
+  setvbuf(stdout, NULL, _IONBF, 0);
+
+// -------------------------------------------------------------------------------
+// Reset Any Previously Aborted Sequences
+// -------------------------------------------------------------------------------
+  temp = 0;
+  lseek(CFG, cntl_reg, SEEK_SET);
+  CHECK( write(CFG,&temp,4) );
+ 
+
+// -------------------------------------------------------------------------------
+// Wait for Flash to be Ready
+// -------------------------------------------------------------------------------
+  lseek(CFG, cntl_reg, SEEK_SET);
+  CHECK( read(CFG,&temp,4) );
+
+  st = time(NULL);
+  lt = st;
+  int ec = 0;
+  int cp = 1;
+
+  while (cp == 1) {
+    lseek(CFG, cntl_reg, SEEK_SET);
+    CHECK( read(CFG,&temp,4) );
+    if ( (temp & 0x80000000) == 0x80000000 ) {
+      cp = 0;
+    }
+    ct = time(NULL);
+    if ((ct - lt) > 5) {
+      printf(".");
+      lt = ct;
+    }
+    if ((ct - st) > 120) {
+      printf ("\nFAILURE --> Flash not ready after 2 min\n");
+      cp = 0;
+      ec = 1;
+    }
+  }
+  printf("\n");
+
+  off_t fsize;
+  struct stat tempstat;
+  int num_blocks;
+  address = 0;  //the user partion.
```

Comment:
Does Semptian have the user partition at address 0? Is there a factory partition at all? If yes, how does it make sure the images switch over? The image build then also requires matching parameters such as BITSTREAM.CONFIG.TIMER_CFG

Reply:
@jsvogt, the Alpha-Data KU115 and Bittware boards also use 0x0 for the user partition.